contracts/libraries/CronUtility_External.sol

```diff
@@ -0,0 +1,565 @@
+// SPDX-License-Identifier: MIT
+
+/*
+  The CronUtility_Internal contract serves two primary functions:
+    * parsing cron-formatted strings like "0 0 * * *" into
+      structs called "Specs"
+    * computing the "next tick" of a cron spec
+
+  Because manipulating strings is gas-expensive in solidity,
+  the intended use of this contract is for users to first convert
+  their cron strings to encoded Spec structs via toEncodedSpec().
+  Then, the user stores the Spec on chain. Finally, users use the nextTick(),
+  function to determine the datetime of the next cron job run.
+
+  Cron jobs are interpreted acording to this fomat:
+
+  ┌───────────── minute (0 - 59)
+  │ ┌───────────── hour (0 - 23)
+  │ │ ┌───────────── day of the month (1 - 31)
+  │ │ │ ┌───────────── month (1 - 12)
+  │ │ │ │ ┌───────────── day of the week (0 - 6) (Monday to Sunday)
+  │ │ │ │ │
+  │ │ │ │ │
+  │ │ │ │ │
+  * * * * *
+
+  Special limitations:
+    * there is no year field
+    * no special characters: ? L W #
+    * lists can have a max length of 26
+    * no words like JAN / FEB or MON / TUES
+*/
+
+pragma solidity 0.8.6;
+
+import "../vendor/Strings.sol";
+import "../vendor/DateTime.sol";
+
+// The fields of a cron spec, by name
+string constant MINUTE = "minute";
+string constant HOUR = "hour";
+string constant DAY = "day";
+string constant MONTH = "month";
+string constant DAY_OF_WEEK = "day of week";
+
+error UnknownFieldType();
+error InvalidSpec(string reason);
+error InvalidField(string field, string reason);
+error ListTooLarge();
+
+// Set of enums representing a cron field type
+enum FieldType {
+  WILD,
+  EXACT,
+  INTERVAL,
+  RANGE,
+  LIST
+}
+
+// A spec represents a cron job by decomposing it into 5 fields
+struct Spec {
+  Field minute;
+  Field hour;
+  Field day;
+  Field month;
+  Field dayOfWeek;
+}
+
+// A field represents a single element in a cron spec. There are 5 types
+// of fields (see above). Not all properties of this struct are present at once.
+struct Field {
+  FieldType fieldType;
+  uint8 singleValue;
+  uint8 interval;
+  uint8 rangeStart;
+  uint8 rangeEnd;
+  uint8 listLength;
+  uint8[26] list;
+}
+
+/**
+ * @title The CronUtility library
+ * @notice A utility contract for encoding/decoding cron strings (ex: 0 0 * * *) into an
+ * abstraction called a Spec. The library also includes a spec function, nextTick(), which
+ * determines the next time a cron job should fire based on the current block timestamp.
+ */
+library CronUtility_External {
```

Comment:
This looks like a direct copy of `CronUtility_Internal`, but with public functions. I think the external version should import the internal version and then wrap the functions with external functions which call the internal library. This would eliminate duplicated code.

Reply:
great call. Done.